src/lib/y2storage/disk_analyzer.rb

```diff
@@ -203,140 +147,143 @@ def candidate_disks
+    # Disks that are suitable for installing Linux.
     #
-    # @return [Array<String>]
-    def scoped_disks
-      @scoped_disks ||= scope == :install_candidates ? candidate_disks : dev_names(all_disks)
+    # @param check_limit [Fixnum] maximum number of disks to check
+    # for installation disks.
+    # @return [Array<Disk>] candidate disks
+    def candidate_disks(check_limit: DEFAULT_CHECK_LIMIT)
+      return @candidate_disks if @candidate_disks
+      @candidate_disks = find_candidate_disks(check_limit)
+      log.info("Found candidate disks: #{@candidate_disks}")
+      @candidate_disks
     end
 
-    # Array with all disks in the devicegraph
+    # Checks if a partition belongs to any VG or to an specefic one.
     #
-    # @return [Array<Disk>]
-    def all_disks
-      devicegraph.disks
+    # @param partition [Partition]
+    # @param vg [LvmVg] volume group
+    def partition_in_vg?(partition, vg = nil)
+      p_vg = partition_vg(partition)
+      return false unless p_vg
+      return true unless vg
+      p_vg == vg
     end
 
-    # List of disks in the devicegraph
+    # Obtains the PV of a partition.
     #
-    # @return [DisksList]
-    def disks
-      devicegraph.disks
+    # @param partition [Partition]
+    # @return [LvmPv, nil]
+    def partition_pv(partition)
+      devicegraph.lvm_pvs.detect { |pv| pv.plain_blk_device == partition }
     end
 
-    # Find disks that look like the current installation medium
-    # (the medium we just booted from to start the installation).
-    #
-    # This is limited with DiskAnalyzer::disk_check_limit because some
-    # architectures (s/390) tend to have a large number of disks, and
-    # checking if a disk is an installation disk involves mounting and
-    # unmounting each partition on that disk.
+    # Obtains the VG of a partition.
     #
-    # @return [Array<String>] device names of installation disks
+    # @param partition [Partition]
+    # @return [LvmVg, nil]
+    def partition_vg(partition)
+      pv = partition_pv(partition)
+      return nil unless pv
+      pv.lvm_vg
+    end
+
+    # Look up devicegraph element by device name.
     #
-    def find_installation_disks
-      usb_disks, non_usb_disks = all_disks.partition { |d| d.usb? }
-      disks = usb_disks + non_usb_disks
+    # @return [Device]
+    def device_by_name(name)
+      Disk.find_by_name(devicegraph, name)
+    end
 
-      if disks.size > @disk_check_limit
-        disks = disks.first(@disk_check_limit)
-        log.info("Installation disk check limit exceeded - only checking #{dev_names(disks)}")
-      end
+  private
+
+    attr_reader :devicegraph
 
-      dev_names(disks.select { |disk| installation_disk?(disk.name) })
+    # Gets data for a set of disks, stores it and returns that data.
+    #
+    # @param *disks [Disk, String] disks to analyze. All disks by default.
+    # @param data [Symbol] data name.
+    # @param limit [Fixnum, nil] maximum number of disks to check.
```

Comment:
The limit should not be in the number of disks to check. Otherwise we get to the consistency API problems I already pointed. It should be in the number of expensive operations to perform.

For example, checking a disk with 10 partitions is more expensive than checking 100 unformatted and unpartitioned disks.

Of course, we would have to rethink the caching algorithm or to make the limit a soft limit,